interop/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala

```diff
@@ -31,25 +54,95 @@ sealed abstract class CatsInstances2 {
     new CatsMonadError[E] with CatsSemigroupK[E] with CatsBifunctor
 }
 
+private class CatsConcurrent extends CatsEffect with Concurrent[Task] {
+  private[this] def toFiber[A](f: Fiber[Throwable, A]): effect.Fiber[Task, A] = new effect.Fiber[Task, A] {
+    override val cancel: Task[Unit] = f.interrupt.void
+
+    override val join: Task[A] = f.join
+  }
+
+  override def liftIO[A](ioa: cats.effect.IO[A]): Task[A] =
+    Concurrent.liftIO(ioa)(this)
+
+  override def cancelable[A](k: (Either[Throwable, A] => Unit) => effect.CancelToken[Task]): Task[A] =
+    IO.async0 { kk: Callback[Throwable, A] =>
+      val token = try {
+        k(e => kk(eitherToExitResult(e)))
+      } catch {
+        case e: Throwable =>
+          throw e
```

Comment:
Hmm, that 

1. would resume asynchronously
2. Allow recovery as a checked exception in case when .async builder failed to import an impure computation due to an exception

I think 2 would be inconsistent with behavior of ZIO's own .async. Not allowing recovery is inconsistent with [some hidden assumptions cats-effect users have wrt being able to catch exceptions in pure code](https://github.com/tpolecat/doobie/issues/800) 
I'd say I just remove this line for now as it's redundant and stay consistent with ZIO semantics.

@alexandru @jdegoes Would it be possible to come up with laws/specs for catching exceptions in thrown in pure code in cats-effect?

Reply:
The assumption that exceptions will be caught in map/flatMap was explicitly [rejected from lawful codification](https://github.com/typelevel/cats-effect/issues/71), and is incompatible with Cats `UIO`, Cats `BIO`, and of course `ZIO`—as well as, arguably, applicative functor laws.

I'd love to make it more clear to library authors building on cats-effect that **any attempt to throw exceptions from pure code passed to higher order methods like map/flatMap will result in undefined behavior**. Libraries like FS2 get this right but I feel the broader ecosystem still assumes `IO(42).map(throw e).attempt === IO(Left(e))`.

This is a problem that may take care of itself as ZIO and other "non catching" effect monads become more widely used—widely enough to generate bug reports in the libraries that currently assume behavior not guaranteed by cats-effect. Also if Cats `UIO` or `BIO` see adoption, that would accelerate the trend further.

The nice thing is with the lossless ZIO error model, we could implement the current expectations (if necessary), but I feel that would not be a good direction. Better to encourage library authors to stop throw exceptions from pure code.